### DISCUSSION TITLE
Update-DbaInstance hanging on completion of patch and reboot server

### DISCUSSION THREAD

**Discussion by salvbing** [dataplat/dbatools#7992] [Q&A]
DBATOOLS RELEASE 1.1.39.
I am trying to execute Update-DbaInstance on a Active/Passive Windows Failover Cluster server environment:
patch the passive node and restart  (SQL SERVER INSTANCE is up on the ACTIVE NODE)
force failover Active to Passive (newly patched server becomes active)
patch the passive node and restart.

Update-DbaInstance gets hung so after 30 minutes I stop the process.  The patch is completed on the server the hang occurred.  I restart the server manually and I get 2 restarts in the event log.   I dont have a lot to go on, Get-DbatoolsLog 

I see it starting process to execute the setup.exe
Then it does  
CallStack    : {at Get-DbatoolsLog<Process>, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.39\allcommands.ps1: 
               line 37261, at <ScriptBlock>, <No file>: line 1}
ComputerName : mycomputer
File         : C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.39\allcommands.ps1
FunctionName : Invoke-Command2
Level        : Debug
Line         : 87953
Message      : Found an existing session 
               dbatools_CredSSP_stuff, reusing it
ModuleName   : dbatools
Runspace     : ca00a1e8-03a8-4b4f-92fd-f35d63cfd2c3
Tags         : {}
TargetObject : 
Timestamp    : 11/29/2021 6:10:44 PM
Type         : Debug
Username     : thatsme

and this gets repeated once after I break and execute the get-dbatoolslog 20 minutes later. 
Any ideas would be helpful.

**Comment by salvbing** [comment]
continuing to get a hang on Update-DbaInstance on Windows Failover Cluster Servers.  Have to issue a Stop-Operation Ctrl+Break
Get-DbatoolLog always shows it is happening in running the setup and not being able to know that it ended?

Starting process [E:\TempUpdate\dbatools_KB5005226_Extract_85542f071d8740ee8f1e185722b3a414\setup.exe] with arguments [/quiet /allinstances /IAcceptSQLServerLicenseTerms] on SERVERB through CredSSP protocol
Found an existing session dbatools_CredSSP_SERVERB, reusing it

**Comment by potatoqualitee** [comment]
@nvarscar  or @andreasjordan can you please take a look? 🙇🏼

**Comment by nvarscar** [comment]
Perhaps try to run the function with `-Debug` switch and see where it's getting stuck? So far this looks like a PSSession issue and I'm not quite sure this has anything to do with the dbatools code

**Comment by salvbing** [comment]
Ok we should have some patching coming up, we will run with DEBUG.  Hope to overcome this HANG issue with reusing existing session.

**Reply by salvbing** [reply]
updated dbatools to 1.1.60.  I am experimenting with -DEBUG SINGLE SERVER, no hangs when using -DEBUG yet and walking through the Debug steps.
DEBUG: [15:36:03][Invoke-Program] Process [E:\TempUpdate\dbatools_KB5007182_Extract_0e241a5893fc441983bf536d2cc774f9\setup.exe] returned exit code 3010
DEBUG: [15:36:32][Invoke-Command2] Found an existing session dbatools_CredSSP_3cc215b3-30c8-4604-ab9b-8ec41fede5af_SQLSERVER2019.domain, reusing it
DEBUG: [15:36:37][Get-DbaCmObject] Configuration loaded | Cache disabled: False
DEBUG: [15:36:39][Get-DbaCmObject] [SQLSERVER2019.domain] Retrieving Management Information
DEBUG: [15:36:40][Get-DbaCmObject] [SQLSERVER2019.domain] Accessing computer using Cim over DCOM
DEBUG: [15:36:42][Get-DbaCmObject] [SQLSERVER2019.domain] Accessing computer using Cim over DCOM - Success
DEBUG: [15:36:43][Invoke-Command2] Found an existing session dbatools_Default_3cc215b3-30c8-4604-ab9b-8ec41fede5af_SQLSERVER2019.domain, reusing it
DEBUG: [15:36:45][Invoke-Command2] Found an existing session dbatools_Default_3cc215b3-30c8-4604-ab9b-8ec41fede5af_SQLSERVER2019.domain, reusing it
DEBUG: [15:36:46][Test-PendingReboot] Reboot pending in the PendingFileRenameOperations registry value
DEBUG: [15:36:48][Invoke-DbaAdvancedUpdate] Restarting computer SQLSERVER2019.domain and waiting for it to come back online

**Comment by salvbing** [comment]
We havent had a hang with -DEBUG, but there have been 2 hangs without -DEBUG.  They have occurred on PASSIVE side of WINDOWS CLUSTER (SQL SERVER not running).  
C:\Program Files\Microsoft SQL Server\150\Setup Bootstrap\Log\Summary.txt indicates the patch succeeded and needs reboot.
the ExtractPath patch unzip is still present.  I reboot the server manually and get these messages from my script:
PS>TerminatingError(New-PSSession): "The running command stopped because the preference variable "ErrorActionPreference" 
or common parameter is set to Stop: [SERVERB.redacted] 
Connecting to remote server SERVERB.redacted failed with the following error message :
The client cannot connect to the destination specified in the request. 
Verify that the service on the destination is running and is accepting requests. 
Consult the logs and documentation for the WS-Management service running on the destination, 
most commonly IIS or WinRM. 
If the destination is the WinRM service, 
run the following command on the destination to analyze and configure the WinRM service: "winrm quickconfig". 
For more information, see the about_Remote_Troubleshooting Help topic."

PS>TerminatingError(): "Connecting to remote server SERVERB.redacted failed with the following error message : 
The client cannot connect to the destination specified in the request. 
Verify that the service on the destination is running and is accepting requests. 
Consult the logs and documentation for the WS-Management service running on the destination, 
most commonly IIS or WinRM. If the destination is the WinRM service, 
run the following command on the destination to analyze and configure the WinRM service: "winrm quickconfig". 
For more information, see the about_Remote_Troubleshooting Help topic."

What additional debugging can we do?

**Reply by nvarscar** [reply]
Do I understand correctly that the messages that you posted are from the terminal where the Update command is hanging and appear there once you have forced the remote server into a restart? And the command is "unstuck" after you restarted the server manually?

**Reply by salvbing** [reply]
yes - "unstuck" after manually restarting target server.
Also got  these messages after restarting the target server on the terminal where the update is hanging first:
WARNING: [17:08:01][Invoke-DbaAdvancedUpdate] Update failed with exit code 
WARNING: [17:08:01][Update-DbaInstance] Update failed: Update failed with exit code 

Checked with Update-DbaInstance after the "unstuck"
WARNING: [17:12:45][Update-DbaInstance] Explicit -Credential might be required when running agains remote hosts and -Path is a network folder
WARNING: [17:12:47][Get-SqlInstanceUpdate] 15.0.4198 on computer [SERVERB.redacted] is already the latest available.
WARNING: [17:12:47][Get-SqlInstanceUpdate] 15.0.4198 on computer [SERVERB.redacted] is already the latest available.

**Comment by salvbing** [comment]
I got one with -DEBUG.  It hung at the [16:51:25][Invoke-Command2] Found an existing session.  I rebooted the server at 17:09. after checking the target C:\Program Files\Microsoft SQL Server\150\Setup Bootstrap\Log\Summary.txt  timestamp 16:54 ( now 17:09)
Overall summary:
  Final result:                  Passed but reboot required, see logs for details
  Exit code (Decimal):           3010
  Start time:                    2022-02-23 16:51:27
  End time:                      2022-02-23 16:54:50
  Requested action:              Patch

Instance MSSQLSERVER overall summary:
  Final result:                  Passed but reboot required, see logs for details
  Exit code (Decimal):           3010
  Start time:                    2022-02-23 16:51:52
  End time:                      2022-02-23 16:54:47
  Requested action:              Patch
I am attaching the transcript.
[Transcript-20220223T1648302693.txt](https://github.com/dataplat/dbatools/files/8134438/Transcript-20220223T1648302693.txt)

**Reply by nvarscar** [reply]
Unfortunately, all I can see is that the Invoke-Command2 (which launches Invoke-Command behind the scenes) didn't get any result code, effectively handing for a whole 1h+ until 17:09 when you restarted the server. This seems like it is an issue with Powershell Remoting which fails to return the results. Found exactly one link with similar symptoms to try: https://stackoverflow.com/questions/29638446/powershell-remote-invocation-mysteriously-hangs

The only thing that can be done here from the dbatools end, as I don't see other options, is to add a timeout, implemented as an optional `-Timeout X` parameter that would be passed to `Invoke-Parallel -RunspaceTimeout X` to limit the wait time for each parallel thread. While this is not a solution per se, at least that would allow to back out from a hanging process.

**Reply by salvbing** [reply]
Do you have any helps in looking at Windows Events or anything on the target server where the Invoke-Command2 is running?
Does -Timeout use  milliseconds?

**Reply by nvarscar** [reply]
Not sure if Windows Events would help you. It's seconds https://github.com/dataplat/dbatools/blob/development/internal/functions/flowcontrol/Invoke-Parallel.ps1#L45

**Reply by nvarscar** [reply]
Sorry, to clarify: this would be a new feature, it doesn't exist just yet.

**Comment by salvbing** [comment]
Can you provide additional explanation of this description in Update-DbaInstance, does CredSSP have something to do with my issue?:  
        When using CredSSP authentication, this function will try to configure CredSSP authentication for PowerShell Remoting sessions.
        If this is not desired (e.g.: CredSSP authentication is managed externally, or is already configured appropriately,)
        it can be disabled by setting the dbatools configuration option 'commands.initialize-credssp.bypass' value to $true.
        To be able to configure CredSSP, the command needs to be run in an elevated PowerShell session.

**Reply by nvarscar** [reply]
I'm not sure if a CredSSP would be causing all the trouble here, as it's merely a communication protocol. However, you can try to work with Kerberos or Default authentication. As long as the path to the exe file is available to the remote server without any extra auth (i.e. local disk access or network path with anonymous access enabled)

**Reply by salvbing** [reply]
Update-DbaInstance does not like -Timeout or -RunSpaceTimeout parameters
Update-DbaInstance : A parameter cannot be found that matches parameter name 'RunspaceTimeout'.
-Authentication options did not get through the -Path network share needed to copy patch to target.
Hang Problem is occurring more often than not.  

VERBOSE: [15:27:54][Invoke-Program] Starting process [E:\TempUpdate\dbatools_KB5008996_Extract_1b2dbaece67d4d4bb6960f577446426d\setup.exe] with arguments [/quie
t /allinstances /IAcceptSQLServerLicenseTerms] on TargetServer.fqdn.com through Credssp protocol

I am Stuck.  This command should not use Credssp protocol - that was only needed for the extract>>>
VERBOSE: [15:27:27][Invoke-DbaAdvancedUpdate] Extracting \\Networklocation\SQLServer\Software\AutoPatch\SQLServer20
19-KB5008996-CU15-x64.exe to E:\TempUpdate\dbatools_KB5008996_Extract_1b2dbaece67d4d4bb6960f577446426d

**Reply by salvbing** [reply]
CallStack    : At Invoke-Command2, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 89318
               	At Invoke-CommandWithFallback, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 89368
               	At Invoke-Program<Process>, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 90234
               	At Invoke-DbaAdvancedUpdate, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 43192
               	At <ScriptBlock>, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 84013
               	At Update-DbaInstance<Process>, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 84017
               	At <ScriptBlock>, E:\Powershell\Update-HITS_Single_ServerVerbose.ps1: Line 56
File         : C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1
FunctionName : Invoke-Command2
Level        : Debug
Line         : 89318
Message      : Found an existing session dbatools_Credssp_3809438b-24a4-429b-873a-ac3012cbbd35_something, reusing it
ModuleName   : dbatools
Runspace     : 3809438b-24a4-429b-873a-ac3012cbbd35
Tags         : {}
TargetObject : 
Timestamp    : 3/3/2022 3:27:54 PM
Type         : Debug

***reboot target manually****

CallStack    : At Invoke-Program<Process>, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 90243
               	At Invoke-DbaAdvancedUpdate, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 43192
               	At <ScriptBlock>, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 84013
               	At Update-DbaInstance<Process>, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1: Line 84017
               	At <ScriptBlock>, E:\Powershell\Update-HITS_Single_ServerVerbose.ps1: Line 56
File         : C:\Program Files\WindowsPowerShell\Modules\dbatools\1.1.74\allcommands.ps1
FunctionName : Invoke-Program
Level        : Debug
Line         : 90243
Message      : Process [E:\TempUpdate\dbatools_KB5008996_Extract_1b2dbaece67d4d4bb6960f577446426d\setup.exe] returned exit code 
ModuleName   : dbatools
Runspace     : 3809438b-24a4-429b-873a-ac3012cbbd35
Tags         : {}
TargetObject : 
Timestamp    : 3/3/2022 4:03:51 PM
Type         : Debug

**Comment by salvbing** [comment]
dbatools 1.1.80
The remote execution through Credssp protocol is hanging after the patch is applied as verified in
C:\Program Files\Microsoft SQL Server\150\Setup Bootstrap\Log\Summary.txt
Overall summary:
Final result: Passed but reboot required, see logs for details
Exit code (Decimal): 3010
Start time: 2022-03-15 13:39:22
End time: 2022-03-15 13:43:35
Requested action: Patch
-verbose output Shows where hang is happening - I manually reboot target MYComputer and Update-DbaInstance proceeds>>>

VERBOSE: [13:39:21][Invoke-Program] Starting process [E:\TempUpdate\dbatools_KB5007182_Extract_6c1a087e89ad413ab250df5ef19f35f0\setup.exe] with arguments [/quiet /allinstances /IAcceptSQLServerLicenseTerms] on MYComputer through Credssp protocol
WARNING: [13:56:36][Invoke-DbaAdvancedUpdate] Update failed with exit code
Confirm
Are you sure you want to perform this action?
Performing the operation "Set variable" on target "Name:
__dbatools_interrupt_function_78Q9VPrM6999g6zo24Qn83m09XF56InEn4hFrA8Fwhu5xJrs6r Value: True".
&Yes Yes to &All &No No to A&ll &Suspend
Y

ComputerName : MyComputer
MajorVersion : 2019
TargetLevel : RTMCU14
KB : 5007182
Successful : False
Restarted : False
InstanceName :
Installer : \AutoPatch\SQLServer2019-KB5007182-CU14-x64.exe
Notes : {Update failed with exit code }

WARNING: [13:56:50][Update-DbaInstance] Update failed: Update failed with exit code

**Reply by salvbing** [reply]
what is the exit code?

VERBOSE: [11:18:06][Invoke-Program] Starting process [E:\TempUpdate\dbatools_KB5010786_Extract_8a51cd036438416e8ed8ef2151b887c5\setup.exe] with arguments [/quiet /allinstances /IAcceptSQLServerLicenseTerms] on UHAIMDBSQA1.umhs.med.umich.edu through Credssp protocol
WARNING: [11:25:27][Invoke-DbaAdvancedUpdate] Update failed with exit code

When went to check the script DBAdvanced up associated with this, we see that it’s calling this code: 
if ($updateResult.ExitCode -eq 3010 -or $restartNeeded)

But we’re not seeing this 3010 code when we’re outputting this to the script which makes it confusing on whether or not we’re actually grabbing this next line. 

I also want to point out that in the Invoke script we see:            $updateResult = Invoke-Program @execParams -Path "$spExtractPath\setup.exe" -ArgumentList $programArgumentList -WorkingDirectory $spExtractPath -Fallback
            $output.ExitCode = $updateResult.ExitCode

**Reply by salvbing** [reply]
So today it is working.  I get this>
Are you sure you want to perform this action?
Performing the operation "Enable the Remote shutdown access rights and restart the computer ...

Somedays it works somedays it hangs.

**Reply by salvbing** [reply]
Removed parameter -ExtractPath.  No more hangs.  However it appears to be executing from the network -PATH location even though as an intial step the Update-DbaInstance indicates that it is copying the patch executable to a directory created on the target computer e.g. C:\dbatools_KB5number_Extract_string.
